src/libmeasurement_kit/ooni/whatsapp.cpp

```diff
@@ -0,0 +1,575 @@
+            bool close_txp = true; // if connected, we must disconnect
+            Entry result = {
+                {"ip", ip},
+                {"port", port},
+                {"status",
+                    {{"success", nullptr},
+                    {"failure", nullptr}}},
+            };
+            if (!!err) {
+                logger->info("tcp failure to %s:%d", ip.c_str(), port);
+                result["status"]["success"] = false;
+                result["status"]["failure"] = err.as_ooni_error();
+                close_txp = false;
+            } else {
+                logger->info("tcp success to %s:%d", ip.c_str(), port);
+                result["status"]["success"] = true;
+                result["status"]["failure"] = false;
+            }
+            (*entry)["tcp_connect"].push_back(result);
+            *ips_tested += 1;
+            if (ips_count == *ips_tested) {
+                if (close_txp == true) {
+                    txp->close([=] { cb(NoError()); });
+                } else {
+                    cb(NoError());
+                }
+            } else {
+                if (close_txp == true) {
+                    // XXX optimistic closure
+                    txp->close([=] {});
+                }
+            }
+        };
+    };
+
+    for (auto const& ip : ips) {
+        // XXX hardcoded
+        std::vector<int> ports {443, 5222};
+        for (auto const& port : ports) {
+            options["host"] = ip;
+            options["port"] = port;
+            options["net/timeout"] = 10.0; //XXX check if set upstream?
+            templates::tcp_connect(options, tcp_cb(ip, port),
+                                   reactor, logger);
+        }
+    }
+    return;
+}
+
+static void dns_many(std::vector<std::string> hostnames,
+                     Var<Entry> entry,
+                     Settings options,
+                     Var<Reactor> reactor,
+                     Var<Logger> logger,
+                     Callback<Error, std::vector<std::string>> cb) {
+    // if ANYthing is consistent, we consider dns not blocked
+    (*entry)["whatsapp_endpoints_status"] = "blocked";
+    (*entry)["whatsapp_dns_inconsistent"] = {};
+    int names_count = hostnames.size();
+    logger->info("whatsapp: %d hostnames", names_count);
+    Var<std::vector<std::string>> good_ips(new std::vector<std::string>);
+    Var<int> names_tested(new int(0));
+    if (names_count == *names_tested) {
+        cb(NoError(), *good_ips);
+    }
+
+
+    auto dns_cb = [=](std::string hostname) {
+        return [=](Error err, Var<dns::Message> message) {
+            if (!!err) {
+                logger->info("whatsapp: dns error for %s", hostname.c_str());
+            } else {
+                std::vector<std::string> this_ips;
+                for (auto answer : message->answers) {
+                    // XXX i don't get this answer format
+                    // copying from web_connectivity
```

Comment:
It's basically that, depending on the response type, the answer is stored in different fields.

How would you like it to be instead?